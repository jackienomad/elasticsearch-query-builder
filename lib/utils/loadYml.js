import jsYml from 'js-yaml'
import { readFileSync, existsSync } from 'fs'

export function loadYml(configFilePath) {
  const formattedFilePath = ensureYmlExt(configFilePath)
  if (!existsSync(formattedFilePath)) {
    throw Error(`Cannot find config: ${formattedFilePath}`)
  }

  return jsYml.safeLoad(readFileSync(formattedFilePath, 'utf-8'))
}

export function ensureYmlExt(configFileName) {
  const extension = configFileName
    .substring(configFileName.length - 4)
    .toLowerCase()
  const suffix = extension !== '.yml' ? '.yml' : ''

  return `${configFileName}${suffix}`
}

export function getPresetFilename(configFileName) {
  const addedSlash = configFileName.charAt(0) !== '/' ? '/' : ''
  return `${__dirname}/config${addedSlash}` + ensureYmlExt(configFileName)
}

export default loadYml
