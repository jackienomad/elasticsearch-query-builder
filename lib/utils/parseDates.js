const { parse, format } = require('date-fns')

const ensureMilliseconds = function(ts) {
  return Math.floor(ts / 10000000000) ? ts : ts * 1000
}

const flatten = function(arrivalDates) {
  if (typeof arrivalDates === 'string') {
    arrivalDates = arrivalDates.split(',')
  } else if (arrivalDates instanceof Array) {
    arrivalDates = arrivalDates.reduce(
      (acc, arrivalDate) => acc.concat(...arrivalDate.split(',')),
      []
    )
  }

  return arrivalDates
}

const mapDates = function(dateArray) {
  return dateArray.map(date => {
    let start = null
    let end = null

    if (date === '' || date === null || date === undefined) {
      throw new Error('Date is blank')
    }

    // check if Number
    if (Number.isNaN(+date)) {
      // check if is a range
      // check if middle character is a dash
      const middle = Math.floor(date.length / 2)
      const middleChar = date.charAt(middle)

      if (middleChar === '-') {
        // is a range
        start = date.substring(0, middle)
        end = date.substring(middle + 1)
      } else {
        start = date
        end = date
      }

      start = Number.isNaN(+start) ? start : ensureMilliseconds(+start)
      end = Number.isNaN(+end) ? end : ensureMilliseconds(+end)
    } else {
      start = ensureMilliseconds(+date)
      end = ensureMilliseconds(+date)
    }

    // standardize format
    // convert start and end to timestamp(UTC) in seconds
    start = parseDateInUTC(start)
    end = parseDateInUTC(end)

    return {
      start,
      end
    }
  })
}

const parseDateInUTC = function(arrivalDate) {
  if (arrivalDate === null || arrivalDate === undefined || arrivalDate === '') {
    throw new Error('Blank arrival date')
  }

  // check if timestamp or not
  if (typeof arrivalDate !== 'number') {
    const timePattern = /.*\d{1,2}:\d{1,2}:\d{1,2}Z?/
    // check if has time included
    if (!timePattern.test(String(arrivalDate))) {
      arrivalDate += ' 00:00:00Z'
    }
    // check if datetime is in UTC format
    if (arrivalDate.substr(-1) !== 'Z') {
      arrivalDate += 'Z'
    }
  }
  return +format(parse(arrivalDate), 'X')
}

const parseDates = function(arrivalDates) {
  // flatten arrival dates into an array
  arrivalDates = flatten(arrivalDates)
  // dates should always be a range
  arrivalDates = mapDates(arrivalDates)

  return arrivalDates
}

module.exports = {
  ensureMilliseconds,
  flatten,
  mapDates,
  parseDates,
  parseDateInUTC
}
