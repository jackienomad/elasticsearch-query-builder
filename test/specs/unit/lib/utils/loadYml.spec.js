import { loadYml, getPresetFilename, ensureYmlExt } from 'qb/@/utils/loadYml'
import path from 'path'

describe('loadYml utility', () => {
  describe('function ensureYmlExt', () => {
    it('should add .yml if missing', () => {
      const input = 'sample'
      const output = ensureYmlExt(input)
      const expectedOutput = 'sample.yml'
      expect(output).toBe(expectedOutput)
    })

    it('should not add .yml if not missing', () => {
      const input = 'sample.yml'
      const output = ensureYmlExt(input)
      const expectedOutput = 'sample.yml'
      expect(output).toBe(expectedOutput)
    })

    it('should not add .yml even if yml extension is in uppercase', () => {
      const input = 'SAMPLE.YML'
      const output = ensureYmlExt(input)
      const expectedOutput = 'SAMPLE.YML'
      expect(output).toBe(expectedOutput)
    })
  })

  describe('function getPresetFilename', () => {
    it('should add slash if missing', () => {
      const input = 'sample.yml'
      const output = getPresetFilename(input)
      const expectedOutput = `/config/${input}`
      const modifiedOutput = output.substring(
        output.length - expectedOutput.length
      )
      expect(modifiedOutput).toBe(expectedOutput)
    })
    it('should not add slash if present', () => {
      const input = '/sample.yml'
      const output = getPresetFilename(input)
      const expectedOutput = `/config/${input.substring(1)}`
      const modifiedOutput = output.substring(
        output.length - expectedOutput.length
      )
      expect(modifiedOutput).toBe(expectedOutput)
    })

    it('should append extension when missing', () => {
      const input = 'sample'
      const output = getPresetFilename(input)
      const expectedOutput = `/config/${input}.yml`
      const modifiedOutput = output.substring(
        output.length - expectedOutput.length
      )
      expect(modifiedOutput).toBe(expectedOutput)
    })
  })

  describe('function loadYml', () => {
    it('it should throw an error if passed config file', () => {
      expect(() => {
        loadYml('non_existent_config')
      }).toThrowError()
    })

    it('should be able to load a valid yml file', () => {
      const ymlpath = path.resolve(__dirname, 'data', 'loadYml.sample.yml')
      const output = loadYml(ymlpath)
      expect(output).toMatchSnapshot()
    })

    it('should be able to load a valid yml file with no specified extension', () => {
      const ymlpath = path.resolve(__dirname, 'data', 'loadYml.sample')
      const output = loadYml(ymlpath)
      expect(output).toMatchSnapshot()
    })
  })
})
