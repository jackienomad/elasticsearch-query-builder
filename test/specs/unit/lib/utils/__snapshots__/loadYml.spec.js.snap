// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadYml utility function loadYml should be able to load a valid yml file 1`] = `
Object {
  "parent1": Object {
    "child1": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
    "child2": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
  },
  "parent2": Object {
    "child1": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
    "child2": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
  },
}
`;

exports[`loadYml utility function loadYml should be able to load a valid yml file with no specified extension 1`] = `
Object {
  "parent1": Object {
    "child1": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
    "child2": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
  },
  "parent2": Object {
    "child1": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
    "child2": Object {
      "prop1": "property1",
      "prop2": "property2",
    },
  },
}
`;
